### PR TITLE
JSON body parameters

### LINKED ISSUE #4: Pass parameters in request body using JSON format
Currently the API accepts using a mixture of urlencoded parameters for endpoints and HTTP basic auth for authentication. This is not optimal because:

* We cannot use both MaaS and Juju authentication for our endpoints
* Parameter passing is a source of confusion

A fine solution going forward would be to switch to accepting parameters from the request body in JSON format.

### PR DESCRIPTION
**Pull request by neoaggelos**
Fixes #4

### CODE REVIEW

File: scripts/juju_inventory_script.py [hunk 1]
```diff
@@ -15,50 +15,51 @@
...
+    url = '{}/juju/inventory'.format(args.server)
+    res = urlopen(
+        Request(url, method='GET'), data=json.dumps(params).encode()).read()
+    print(res.decode())
 
-            inventory = json.loads(res.decode('utf-8'))
-            print(json.dumps(inventory, indent=4))
-            sys.exit(0)
 
-        elif sys.argv[1] == '--host':
```
Comment (skatsaounis) [inline scripts/juju_inventory_script.py]: Is there a reason to remove this check?

File: scripts/juju_inventory_script.py [hunk 2]
```diff
@@ -15,50 +15,51 @@
...
 import json
 import os
-import sys
-import urllib.request
+from urllib.request import urlopen, Request
 
-inventory_url = 'http://hostname:5000/juju/inventory'
 
+def juju_inventory(args):
+    """Fetches the Juju inventory from the GRNet Ansible Inventory Server
```
Comment (skatsaounis) [inline scripts/juju_inventory_script.py]: GRNet should be removed

File: scripts/juju_inventory_script.py [hunk 3]
```diff
@@ -15,50 +15,51 @@
...
-            print(json.dumps(inventory, indent=4))
-            sys.exit(0)
 
-        elif sys.argv[1] == '--host':
-            # hostvars not supported yet, exit quickly to minimize the lookup
-            # cost
-            print(json.dumps({}))
-            sys.exit(0)
+def main():
+    parser = argparse.ArgumentParser(
```
Comment (skatsaounis) [inline scripts/juju_inventory_script.py]: Is there a reason for moving to argument parser rather than reading environment variables?

File: scripts/maas_inventory_script.py [hunk 4]
```diff
@@ -15,47 +15,42 @@
...
 import json
 import os
-import sys
-import urllib.request
+from urllib.request import urlopen, Request
 
-inventory_url = 'http://hostname:5000/maas/inventory'
+
+def maas_inventory(args):
+    """Fetches the MaaS inventory from the GRNet Ansible Inventory Server
```
Comment (skatsaounis) [inline scripts/maas_inventory_script.py]: GRNet should be removed

File: scripts/maas_inventory_script.py [hunk 5]
```diff
@@ -15,47 +15,42 @@
...
-            req.add_header(
-                'Authorization',
-                'Basic {}'.format(encoded_credentials.decode('ascii')))
-            res = urllib.request.urlopen(req).read()
-
-            inventory = json.loads(res.decode('utf-8'))
-            print(json.dumps(inventory, indent=4))
-            sys.exit(0)
-
-        elif sys.argv[1] == '--host':
```
Comment (skatsaounis) [inline scripts/maas_inventory_script.py]: same with the previous inventory script. I see no reason to remove this check

File: scripts/maas_inventory_script.py [hunk 6]
```diff
@@ -15,47 +15,42 @@
...
-            print(json.dumps({}))
-            sys.exit(0)
-
-        else:
-            raise Exception(
-                'Unknown argument. Please use either --list or --host')
-    except Exception as e:
-        print('Error: {}'.format(str(e)), file=sys.stderr)
-        sys.exit(1)
+    parser = argparse.ArgumentParser(
```
Comment (skatsaounis) [inline scripts/maas_inventory_script.py]: same with the previous script. These files are meant to be called by ansible-inventory binary. Thus, there is no reason to use parse because they are not purposed to be used directly from a user. In case user wants to call the inventory server then he has to call REST api directly.

File: ansible_inventory_server/maasrest.py [hunk 7]
```diff
@@ -16,47 +16,82 @@
 from collections import defaultdict
 import json
 
-from maas.client import connect
+from maas.client.bones import SessionAPI
```
Comment (skatsaounis) [inline ansible_inventory_server/maasrest.py]: this change is irrelevant with this PR IMO

File: ansible_inventory_server/maasrest.py [hunk 8]
```diff
@@ -16,47 +16,82 @@
...
 
 from ansible_inventory_server.utils import ApiRequestHandler
-from ansible_inventory_server import settings
+
+
+__machines = None
+
+
+def filter_maas_machine_info(machine):
```
Comment (skatsaounis) [inline ansible_inventory_server/maasrest.py]: This change is irrelevant with this PR IMO

File: ansible_inventory_server/jujurest.py [hunk 9]
```diff
@@ -14,119 +14,101 @@
...
+                'model_uuid', settings.MODEL_UUID),
+            endpoint=parameters.get('juju', {}).get(
+                'endpoint', settings.JUJU_ENDPOINT)
+        )
+        return model
+    except:
+        return None
+
+
+async def get_juju_status(parameters):
```
Comment (skatsaounis) [inline ansible_inventory_server/jujurest.py]: The refactoring here is not part of this PR

File: ansible_inventory_server/jujurest.py [hunk 10]
```diff
@@ -14,119 +14,101 @@
...
+    """Connects to a Juju model and returns status"""
+    model = await get_juju_model(parameters)
+    if model is None:
+        return None
+
+    status = await model.get_status(parameters.get('juju', {}).get('filters'))
+    return status
+
+
+def juju_filter_machine_info(machine, data):
```
Comment (skatsaounis) [inline ansible_inventory_server/jujurest.py]: Same here. This refactor has nothing to do with JSON body parameters

File: ansible_inventory_server/jujurest.py [hunk 11]
```diff
@@ -14,119 +14,101 @@
...
+            result[lxd] = juju_filter_machine_info(
+                lxd, lxd_data)
 
+            result[lxd]['parent'] = machine
+            result[machine]['containers'].append(lxd)
 
-def inventory_host_units(status):
-    """returns an {'ip_address': 'unit_name'} dict"""
-    result = defaultdict(lambda: [])
+    # update list of Juju apps and subordinate units
```
Comment (skatsaounis) [inline ansible_inventory_server/jujurest.py]: same here. This refactor is irrelevant

File: ansible_inventory_server/jujurest.py [hunk 12]
```diff
@@ -135,39 +117,51 @@ def create_response(self, status):
...
-        machines = get_machines_ips(status)
-        return to_inventory_object(status, machines)
+        machines = get_juju_machines(status)
+
+        result = {
+            '_meta': {'hostvars': {}},
+            status.model.name: {'children': []}
+        }
+
+        for machine, machine_data in machines.items():
```
Comment (skatsaounis) [inline ansible_inventory_server/jujurest.py]: This refactor is irrelevant with this PR

File: ansible_inventory_server/jujurest.py [hunk 13]
```diff
@@ -135,39 +117,51 @@ def create_response(self, status):
...
-        for app_name, app_data in status['applications'].items():
+        all_machines = get_juju_machines(status)
 
-            app_has_nrpe = 'nrpe-external-master' in app_data['relations']
-            for unit_name, unit_data in (app_data.get('units') or {}).items():
-                machine_id = unit_data['machine']
-                unit_has_nrpe = any(
-                    x.startswith('nrpe-')
-                    for x in (unit_data.get('subordinates') or []))
+        # for units with a `nrpe-external-master` relation, machine has Juju
```
Comment (skatsaounis) [inline ansible_inventory_server/jujurest.py]: same here.

**Review by skatsaounis — CHANGES_REQUESTED** [review]
Please move any refactor present to this PR to a new one and fix some minor text.

File: scripts/juju_inventory_script.py [hunk 14]
```diff
@@ -30,18 +27,18 @@ def main():
...
+            params = {
+                'juju': {
+                    'username': os.getenv('JUJU_USERNAME'),
+                    'password': os.getenv('JUJU_PASSWORD'),
+                    'model_uuid': os.getenv('JUJU_MODEL_UUID'),
+                    'cacert': os.getenv('JUJU_CACERT').replace(r'\n', '\n'),
+                    'endpoint': os.getenv('JUJU_ENDPOINT')
+                }
+            }
+            req = Request('{}/juju/inventory'.format(os.getenv('AIS_URL')),
```
Comment (skatsaounis) [inline scripts/juju_inventory_script.py:48]: Introduction of AIS_URL should be documented IMO because we are expecting user to provide a value

**Review by skatsaounis — CHANGES_REQUESTED** [review]
Please update README (Usage part) and I think we are ready

File: scripts/juju_inventory_script.py (hunk 14 above)
Comment (neoaggelos) [reply scripts/juju_inventory_script.py:48]: Updated docs regarding the API changes. Documentation is in dire need of a re-write, but this should be done on a separate pull request

File: ansible_inventory_server/jujurest.py [hunk 15]
```diff
@@ -23,25 +23,33 @@
...
+            username=parameters['juju']['username'],
+            password=parameters['juju']['password'],
+            uuid=parameters.get('juju', {}).get(
+                'model_uuid', settings.MODEL_UUID),
+            endpoint=parameters.get('juju', {}).get(
+                'endpoint', settings.JUJU_ENDPOINT)
+        )
+        status = await model.get_status(
+            parameters.get('juju', {}).get('filters'))
+
```
Comment (skatsaounis) [inline ansible_inventory_server/jujurest.py:48]: please add `await model.disconnect()` as it was missing

**Review by skatsaounis — CHANGES_REQUESTED** [review]

File: ansible_inventory_server/jujurest.py (hunk 15 above)
Comment (neoaggelos) [reply ansible_inventory_server/jujurest.py:48]: used `asyncio.ensure_future()` instead. We do not have to wait for model to disconnect to finish processing the request

**Review by skatsaounis — APPROVED** [review]
